vshard/storage/init.lua

```diff
@@ -3229,6 +3255,74 @@ local function storage_call(bucket_id, mode, name, args)
     return ok, ret1, ret2, ret3
 end
 
+-- Multicall wrapper
+-- Refer non-moved buckets, call user's function, unrefer buckets
+-- and return results with a list of moved buckets.
+local function storage_multicall(bucket_ids, mode, func, args)
+    local ok_ref, res, err
+    local ref_buckets = {}
+    local bucket
+    local moved_buckets = {}
+    local prev_bucket_id = nil
+
+    -- Sort bucket ids to detect duplicates.
+    table.sort(bucket_ids)
+    for _, bucket_id in pairs(bucket_ids) do
+        -- Avoid double referencing of the same bucket.
+        if bucket_id == prev_bucket_id then
+            goto continue
+        end
+
+        -- Collect moved buckets.
+        bucket = box.space._bucket:get({bucket_id})
+        if not bucket then
+            table.insert(moved_buckets, bucket_id)
+            goto continue
+        end
+
+        -- Reference the bucket.
+        ok_ref, err = bucket_ref(bucket_id, mode)
+        if not ok_ref then
+            err = lerror.make(err)
+            goto fail
+        end
+        table.insert(ref_buckets, bucket_id)
```

Comment:
You don't need `ref_buckets`. When you meet a first error, you can iterate `bucket_ids` again to the same position and do all the unrefs, and return error. But in case of success `ref_buckets` is equal to `bucket_ids`. Which means you don't need it.